_source/_posts/2024-04-30-spring-api-fga.md

```diff
@@ -0,0 +1,901 @@
+      "quotaBytesUsed":null,
+      "version":null,
+      "originalFilename":null,
+      "fileExtension":".txt"
+   }
+]
+```
+
+## Define an authorization model
+
+At a high level, an authorization model is defined by indicating user types, object types, and relationships between them. As we are not going to deep-dive on [ReBAC](https://openfga.dev/docs/authorization-concepts#what-is-relationship-based-access-control) in this guide, you can refer to OpenFGA documentation for learning about modeling concepts. Under the [Advanced use-cases](https://openfga.dev/docs/modeling/advanced) section in the doc, there is a simplified authorization model for a [Google Drive](https://openfga.dev/docs/modeling/advanced/gdrive) application ready to test:
+
+```fga
+model
+  schema 1.1
+
+type user
+
+type document
+  relations
+    define owner: [user, domain#member] or owner from parent
+    define writer: [user, domain#member] or owner or writer from parent
+    define commenter: [user, domain#member] or writer or commenter from parent
+    define viewer: [user, user:*, domain#member] or commenter or viewer from parent
+    define parent: [document]
+
+type domain
+  relations
+    define member: [user]
+```
+
+The graph below is a model preview generated by the [OpenFGA playground](https://openfga.dev/docs/getting-started/setup-openfga/playground) where you can see the [direct and implied](https://openfga.dev/docs/concepts#what-are-direct-and-implied-relationships) relations of the model.
+
+{% img blog/spring-api-fga/fga-model.png alt:"Authorization model preview" width:"700" %}{: .center-image }
+
+The OpenFGA authorization model is expressed in a configuration language that can be presented in DSL or JSON syntax. The JSON syntax is accepted by the API. The DSL adds syntactic sugar on top of JSON for ease of use and can be transformed to JSON using the FGA CLI. The syntax above is the document authorization model in DSL. Copy the model DSL from above to the file `src/main/resources/fga/auth-model.fga` file, install the [FGA CLI](https://openfga.dev/docs/getting-started/cli), and transform the model to JSON:
+
+
+```shell
+fga model transform --file=auth-model.fga > auth-model.json
+```
+
+The transformed model should be like the JSON below:
+
+```json
+// auth-model.json
+{
+   "schema_version":"1.1",
+   "type_definitions":[
+      {
+         "type":"user"
+      },
+      {
+         "metadata":{...},
+         "relations":{...},
+         "type":"document"
+      },
+      {
+         "metadata":{...},
+         "relations":{...},
+         "type":"domain"
+      }
+   ]
+}
+```
+
+Place the file at `src/main/resources/fga/auth-model.json` as it will be required later.
+
+## Add fine-grained authorization (FGA) with OpenFGA
+
+Now let's integrate fine-grained authorization into the application. The OpenFGA team has just released the OpenFGA Spring Boot Starter 0.0.1, and you can add it to your project with the following dependency:
```

Comment:
Please link to the github repo of the starter

Reply:
Added links to the repo of the starter